src/interface/data_collection.hpp

```diff
@@ -44,9 +51,16 @@ class DataCollection {
 
   void SetMeshPointer(Mesh *pmesh) { pmy_mesh_ = pmesh; }
 
-  template <typename ID_t>
-  std::shared_ptr<T> &Add(const std::string &name, const std::shared_ptr<T> &src,
+  template <class SRC_t, typename ID_t>
+  std::shared_ptr<T> &Add(const std::string &name, const std::shared_ptr<SRC_t> &src,
                           const std::vector<ID_t> &fields, const bool shallow) {
+    if constexpr (!((std::is_same_v<SRC_t, MeshBlock> &&
+                     std::is_same_v<T, MeshBlockData<Real>>) ||
+                    std::is_same_v<SRC_t, T>)) {
+      // SRC_t and T are incompatible
+      static_assert(always_false<SRC_t>);
```

Comment:
Why not just `static_assert` the boolean in the if constexpr and add the comment as a string?

Reply:
see above. String is probably useful though.

Reply:
string added